src/Microsoft.AspNetCore.Mvc.Core/Formatters/TextInputFormatter.cs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Microsoft.AspNetCore.Mvc.Formatters
+{
+    /// <summary>
+    /// Reads an object from the request body.
+    /// </summary>
+    public abstract class TextInputFormatter : InputFormatter
+    {
+        /// <summary>
+        /// Returns UTF8 Encoding without BOM and throws on invalid bytes.
+        /// </summary>
+        protected static readonly Encoding UTF8EncodingWithoutBOM
+            = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
+
+        /// <summary>
+        /// Returns UTF16 Encoding which uses littleEndian byte order with BOM and throws on invalid bytes.
+        /// </summary>
+        protected static readonly Encoding UTF16EncodingLittleEndian
+            = new UnicodeEncoding(bigEndian: false, byteOrderMark: true, throwOnInvalidBytes: true);
+
+        /// <summary>
+        /// Gets the mutable collection of character encodings supported by
+        /// this <see cref="TextInputFormatter"/>. The encodings are
+        /// used when reading the data.
+        /// </summary>
+        public IList<Encoding> SupportedEncodings { get; } = new List<Encoding>();
+
+        /// <inheritdoc />
+        public override Task<InputFormatterResult> ReadAsync(InputFormatterContext context)
+        {
+            var request = context.HttpContext.Request;
+            if (request.ContentLength == 0)
+            {
+                return InputFormatterResult.SuccessAsync(GetDefaultValueForType(context.ModelType));
+            }
+
+            var requestEncoding = GetRequestEncoding(request);
+            var selectedEncoding = SelectCharacterEncoding(context, requestEncoding);
+            if (requestEncoding != null && selectedEncoding == null && SupportedEncodings.Count > 0)
```

Comment:
Let's move this into `SelectCharacterEncoding`. Add the error if `SelectCharacterEncoding` returns null.